folding-schemes/src/frontend/circom/utils.rs

```diff
@@ -0,0 +1,146 @@
+}
+
+impl<F: PrimeField> CircomWrapper<F> {
+    // Creates a new instance of the CircomWrapper with the file paths.
+    pub fn new(r1cs_filepath: PathBuf, wasm_filepath: PathBuf) -> Self {
+        CircomWrapper {
+            r1cs_filepath,
+            wasm_filepath,
+            _marker: PhantomData,
+        }
+    }
+
+    // Aggregated function to obtain R1CS and witness from Circom.
+    pub fn extract_r1cs_and_witness(
+        &self,
+        inputs: &[(String, Vec<BigInt>)],
+    ) -> Result<(R1CS<F>, Option<Vec<F>>), Error> {
+        // Extracts the R1CS
+        let file = File::open(&self.r1cs_filepath)?;
+        let reader = BufReader::new(file);
+        let r1cs_file = r1cs_reader::R1CSFile::<F>::new(reader)?;
+        let r1cs = r1cs_reader::R1CS::<F>::from(r1cs_file);
+
+        // Extracts the witness vector
+        let witness_vec = self.extract_witness(inputs)?;
+
+        Ok((r1cs, Some(witness_vec)))
+    }
+
+    // Extracts the witness vector as a vector of PrimeField elements.
+    pub fn extract_witness(&self, inputs: &[(String, Vec<BigInt>)]) -> Result<Vec<F>, Error> {
+        let witness_bigint = self.calculate_witness(inputs)?;
+
+        witness_bigint
+            .iter()
+            .map(|big_int| {
+                self.num_bigint_to_ark_bigint(big_int)
+                    .and_then(|ark_big_int| {
+                        F::from_bigint(ark_big_int)
+                            .ok_or_else(|| Error::Other("could not get F from bigint".to_string()))
+                    })
+            })
+            .collect()
+    }
+
+    // Calculates the witness given the Wasm filepath and inputs.
+    pub fn calculate_witness(
+        &self,
+        inputs: &[(String, Vec<BigInt>)],
+    ) -> Result<Vec<BigInt>, Error> {
+        let mut calculator = WitnessCalculator::new(&self.wasm_filepath)
+            .map_err(|_| Error::Other("Could not create new WitnessCalculator".to_string()))?;
+        calculator
+            .calculate_witness(inputs.iter().cloned(), true)
+            .map_err(|_| Error::Other("Failed to calculate witness".to_string()))
+    }
+
+    // Converts a num_bigint::BigInt to a PrimeField::BigInt.
+    pub fn num_bigint_to_ark_bigint(&self, value: &BigInt) -> Result<F::BigInt, Error> {
+        let big_uint = value
+            .to_biguint()
+            .ok_or(Error::Other("BigInt is negative".to_string()))?;
+        F::BigInt::try_from(big_uint).map_err(|_| {
+            Error::Other("Failed to convert num_bigint::BigInt to PrimeField::BigInt".to_string())
+        })
+    }
+
+    // Converts a PrimeField element to a num_bigint::BigInt representation.
+    pub fn ark_primefield_to_num_bigint(&self, value: F) -> BigInt {
+        let primefield_bigint: F::BigInt = value.into_bigint();
+        let bytes = primefield_bigint.to_bytes_be();
+        BigInt::from_bytes_be(Sign::Plus, &bytes)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use ark_bn254::Fr;
```

Comment:
In step_constraints and wrapper circuit, both `ark_pallas` and `ark_bn254` work, but in circom's `Fcircuit` using `extract_r1cs_and_witness`, only `ark_bn254` works. Does this cause problems from an efficiency or something point? Or is this still ok? @arnaucube

Reply:
Maybe for this PR we can leave it as it is with just BN254 support (the default in Circom and the only one supported at the current version of Sonobe for the Decider (DeciderEth)), and it can be something to do in the future

Reply:
Ok, closing it and adding this to a list of things to keep track of!

Reply:
I understood. Thank you!